pyanaconda/display.py

```diff
@@ -0,0 +1,328 @@
+        flags.vncquestion = False
+        for error_message in vnc_error_messages:
+            stdout_log.warning(error_message)
+
+    # Should we try to start Xorg?
+    want_x = anaconda.gui_mode and not (flags.preexisting_x11 or flags.usevnc)
+
+    # X on a headless (e.g. s390) system? Nonsense!
+    if want_x and anaconda.isHeadless:
+        stdout_log.warning(_("DISPLAY variable not set. Starting text mode."))
+        anaconda.display_mode = constants.DISPLAY_MODE_TUI
+        anaconda.gui_startup_failed = True
+        time.sleep(2)
+        want_x = False
+
+    # Is Xorg is actually available?
+    if want_x and not os.access("/usr/bin/Xorg", os.X_OK):
+        stdout_log.warning(_("Graphical installation is not available. "
+                             "Starting text mode."))
+        time.sleep(2)
+        anaconda.display_mode = constants.DISPLAY_MODE_TUI
+        want_x = False
+
+    if anaconda.tui_mode and flags.vncquestion:
+        #we prefer vnc over text mode, so ask about that
+        message = _("Text mode provides a limited set of installation "
+                    "options. It does not offer custom partitioning for "
+                    "full control over the disk layout. Would you like "
+                    "to use VNC mode instead?")
+        ask_vnc_question(anaconda, vnc_server, message)
+        if not anaconda.ksdata.vnc.enabled:
+            # user has explicitly specified text mode
+            flags.vncquestion = False
+
+    display_mode_name = constants.DISPLAY_MODE_NAMES.get(anaconda.display_mode)
+    if display_mode_name:
+        log.info("Display mode = %s", anaconda.display_mode)
+    elif anaconda.display_mode:
+        log.error("Unknown display mode: %s", anaconda.display_mode)
+    else:
+        log.error("Display mode not set!")
+
+    startup_utils.check_memory(anaconda, options)
+
+    # check_memory may have changed the display mode
+    want_x = want_x and (anaconda.gui_mode)
+    if want_x:
+        try:
+            start_x11()
+            do_startup_x11_actions()
+        except (OSError, RuntimeError) as e:
+            log.warning("X startup failed: %s", e)
+            stdout_log.warning("X startup failed, falling back to text mode")
+            anaconda.display_mode = constants.DISPLAY_MODE_TUI
+            anaconda.gui_startup_failed = True
+            time.sleep(2)
+
+        if not anaconda.gui_startup_failed:
+            do_extra_x11_actions(options.runres, gui_mode=anaconda.gui_mode)
+
+    if anaconda.tui_mode and anaconda.gui_startup_failed and flags.vncquestion and not anaconda.ksdata.vnc.enabled:
+        message = _("X was unable to start on your "
+                    "machine.  Would you like to "
+                    "start VNC to connect to "
+                    "this computer from another "
+                    "computer and perform a "
+                    "graphical installation or continue "
+                    "with a text mode installation?")
```

Comment:
Yeah, it could be there for historic reasons, but it will be a single string anyway when the TUI widget gets it. So I've merged it to just 3 lines.